pkg/auth/handler/sso/auth_handler.go

```diff
@@ -0,0 +1,304 @@
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	PasswordAuthProvider    password.Provider           `dependency:"PasswordAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderFactory         *sso.ProviderFactory        `dependency:"SSOProviderFactory"`
+	Provider                sso.Provider
+	SSOSetting              sso.Setting
+	ProviderName            string
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// reference from APIHandlerToHandler
+	var err error
+
+	defer func() {
+		if err != nil {
+			e := skyerr.MakeError(err)
+			statusCode := nextSkyerr.ErrorDefaultStatusCode(e)
+			http.Error(rw, err.Error(), statusCode)
+		}
+	}()
+
+	if h.Provider == nil {
+		err = skyerr.NewInvalidArgument("Provider is not supported", []string{h.ProviderName})
+		return
+	}
+
+	payload, err := h.DecodeRequest(r)
+	if err != nil {
+		return
+	}
+
+	if err = payload.Validate(); err != nil {
+		return
+	}
+
+	if err := h.TxContext.BeginTx(); err != nil {
+		panic(err)
+	}
+
+	defer func() {
+		if err != nil {
+			h.TxContext.RollbackTx()
+		} else {
+			h.TxContext.CommitTx()
+		}
+	}()
+
+	reqPayload := payload.(AuthRequestPayload)
+	oauthAuthInfo, err := h.getAuthInfo(reqPayload)
+	if err != nil {
+		return
+	}
+	resp, err := h.getResp(oauthAuthInfo)
+	if err != nil {
+		return
+	}
+
+	// wrap resp in result
+	resp = handler.APIResponse{
+		Result: resp,
+	}
+
+	// handle callback url by ux_mode
+	UXMode := oauthAuthInfo.State.UXMode
+	callbackURL := oauthAuthInfo.State.CallbackURL
+	allowedCallbackURLs := h.SSOSetting.AllowedCallbackURLs
+	err = h.validateCallbackURL(allowedCallbackURLs, callbackURL)
+	if err != nil {
+		return
+	}
+
+	// handle authResp by UXMode
+	switch oauthAuthInfo.State.UXMode {
+	case sso.WebRedirect.String(), sso.WebPopup.String():
+		err = h.handleSessionResp(rw, r, UXMode, callbackURL, resp)
+	case sso.IOS.String(), sso.Android.String():
+		err = h.handleRedirectResp(rw, r, UXMode, callbackURL, resp)
```

Comment:
@kaojohnny I double checked the code of sso plugin, `handleSessionResp` and `handleRedirectResp` should have ability to handle some of the errors(e.g. invalid code, client.. etc, see `callback_redirect_response` decorator in sso plugin).

We should print the error to the response body only for some critical case, e.g. callback url is not valid we should not redirect the result to the provided callback url. Or any other case we cannot redirect the result.

Redirect the result (both success and error) back to the app is important, so that user can go back to the original app. SDK will parse the result and call the success or error callback accordingly.

Reply:
1. The callback url validation handling is correct in this PR.

2. `callback_redirect_response ` is a decorator used by `auth_handler`. For all skygear errors raise in `auth_handler`, `callback_redirect_response` will catch it and change it to response object in Line 124. `handler_redirect_response` and `handler_session_response` will encode the error response, set it in cookies and redirect or render html.

The other parts of the PR looks good and the PR becomes huge, maybe let me merge it first and open another PR for the error handling?

Feel free ping me if we need discuss directly. Thanks Johnny!

Reply:
Thanks, if you think it is okay, please help merge it first, I will have another PR to handle errors.